pkg/adaptors/mutex/mutex.go

```diff
@@ -0,0 +1,95 @@
+package mutex
+
+import (
+	"context"
+	"fmt"
+	"github.com/alexflint/go-filemutex"
+	"github.com/google/wire"
+	"github.com/int128/kubelogin/pkg/adaptors/logger"
+	"golang.org/x/xerrors"
+	"os"
+	"path"
+)
+
+//go:generate mockgen -destination mock_mutex/mock_mutex.go github.com/int128/kubelogin/pkg/adaptors/mutex Interface
+
+var Set = wire.NewSet(
+	wire.Struct(new(Mutex), "*"),
+	wire.Bind(new(Interface), new(*Mutex)),
+)
+
+type Interface interface {
+	Acquire(ctx context.Context, name string) (*Lock, error)
+	Release(lock *Lock) error
+}
+
+// Lock holds the lock data.
+type Lock struct {
+	Data interface{}
+	Name string
+}
+
+type Mutex struct {
+	Logger logger.Interface
+}
+
+// internalAcquire wait for acquisition of the lock
+func internalAcquire(fm *filemutex.FileMutex) chan error {
+	result := make(chan error)
+	go func() {
+		if err := fm.Lock(); err != nil {
+			result <- err
+		}
+		close(result)
+	}()
+	return result
+}
+
+// closeAndDelete disposes of resources associated with a lock
+func closeAndDelete(fm *filemutex.FileMutex, lfn string, log logger.Interface) {
+	if err := fm.Close(); err != nil {
+		log.V(1).Infof("Error closing lock file %s: %s", lfn, err)
+	}
+	if err := os.Remove(lfn); err != nil {
+		log.V(1).Infof("error deleting lock file %s: %s", lfn, err)
+	}
+}
+
+// LockFileName get the lock file name from the lock name.
+func LockFileName(name string) string {
+	return path.Join(os.TempDir(), fmt.Sprintf(".kubelogin.%s.lock", name))
+}
+
+// Acquire acquire a lock for the specified name. The context could be used to set a timeout.
+func (m *Mutex) Acquire(ctx context.Context, name string) (*Lock, error) {
+	lfn := LockFileName(name)
+	fm, err := filemutex.New(lfn)
+	if err != nil {
+		return nil, xerrors.Errorf("error creating mutex file %s: %w", lfn, err)
+	}
+
+	lockChan := internalAcquire(fm)
+	select {
+	case <-ctx.Done():
+		closeAndDelete(fm, lfn, m.Logger)
+		return nil, ctx.Err()
+	case err := <-lockChan:
+		if err != nil {
+			err = xerrors.Errorf("error acquiring lock on file %s: %w", lfn, err)
+			closeAndDelete(fm, lfn, m.Logger)
```

Comment:
Do you have some reason to return `Lock` value even if error occurred? I think the following code is better for understanding the error behavior.

```suggestion
			closeAndDelete(fm, lfn, m.Logger)
			return nil, xerrors.Errorf("error acquiring lock on file %s: %w", lfn, err)
```

Reply:
No reason. I fixed it.